Implement getbestblockhash

ala-mode opened the pull request:
Fixes #198

ala-mode committed:
begin getbestblockhash, BlockExplorerIndexer trait, backends state fn

idky137 committed:
added chain_index mod, added ChainBlock struct

idky137 committed:
fixed cyphertext length

idky137 committed:
added TryFrom<Transaction> for TxData

idky137 committed:
added tryfrom fullblock + aux data for ChainBlock

idky137 committed:
fixed TxOut deserialisation

idky137 committed:
added into compact for block and transactions

idky137 committed:
added dcbor encoding to types.rs

idky137 committed:
Merge branch 'dev' into finalised_state_2_0

gustavovalverde committed:
refactor(config)!: migrate to figment for layered configuration parsing

Implement a configuration system using Figment to support a layered approach
(defaults → TOML → environment variables).

Replace the custom TOML parsing logic with standard Serde deserialization
and structured configuration extraction. Add integration tests with the Jail testing utility
for better isolation and coverage of configuration scenarios.

BREAKING CHANGE:
Optional fields in `zindexer.toml` that are not strings or paths (e.g.,
`Option<usize>`, `Option<bool>`) and were previously set to the string
literal "None" will now cause a parsing error. These keys should be
omitted from the TOML to be interpreted as `None`.
Optional string or path fields set to "None" will now be interpreted as
the literal string/path "None", rather than `std::option::Option::None`.
Omit these keys for them to be `None`.

idky137 committed:
implemented authdataroot

idky137 committed:
Merge branch 'finalised_state_2_0' of github.com:idky137/zaino into finalised_state_2_0

idky137 committed:
Merge branch 'dev' of github.com:zingolabs/zaino into finalised_state_2_0

AloeareV committed:
better zebra error propagation

zancas commented on zaino-state/src/indexer.rs:
```diff
@@ -412,6 +411,15 @@ pub trait ZcashIndexer: Send + Sync + 'static {
+#[async_trait]
+pub trait BlockExplorerIndexer: Send + Sync + Clone + ZcashIndexer + 'static {
+    /// Return best block (tip) hash
+    async fn get_best_block_hash(&self) -> Result<String, Self::Error>;
```
I think it should be:

`get_best_blockhash`

dorianvp committed:
Merge branch 'dev' into implement_getbestblockhash

nachog00 committed:
runs-on: arc-sh-runners

nachog00 committed:
workflow_dispatch

nachog00 committed:
on.push.branches= [test-sh-cache]

nachog00 committed:
all use arc-sh-runners and minio cache

nachog00 committed:
added port env var and spec on uses

nachog00 committed:
artifacts now handled by cache action

nachog00 committed:
improved cacheing keys
In this commit, the CI workflow file (`ci.yml`) for the GitHub Actions setup has been updated to improve the caching mechanism used for Rust builds. Here are the key changes made:

1. **Separated Cache Steps**: The caching logic for cargo dependencies has been separated into distinct restore and save steps, enhancing clarity and modularity.

2. **Updated Key Naming**: Cache keys have been updated to use `cargo-deps-${{ hashFiles('**/Cargo.lock') }}`. This change ensures that cache hits are more likely to occur only when dependencies change, leveraging the Cargo lock file for version accuracy.

3. **Conditional Cache Saving**: The save step for cargo dependencies is now conditional. It will only execute if the restore step has determined there's no cache hit, ensuring unnecessarysaving operations are avoided.

4. **Scoped Cache Directory**: Specifically caching `target/debug/deps` rather than the entire `target` directory focuses caching efforts on dependencies, which should help speed up buildsby restoring only what's needed.

nachog00 committed:
empty commit to trigger another run of CI

dorianvp committed:
feat(validator): implement `getblockcount`

dorianvp committed:
chore: run clippy

AloeareV committed:
adress clippy lints, box some error variants

dorianvp committed:
chore: remove todos

ala-mode committed:
Merge branch 'dev' into implement_getbestblockhash

ala-mode committed:
adjust fn naming

dorianvp committed:
Merge branch 'dev' into implement-getblockcount

nachog00 committed:
Add new CI Docker setup and remove old one
- Added a new `.env.ci-dep-versions` file to specify dependency versions for CI
- Created `Dockerfile.ci` to build a Docker image with specific versions of zcashd, lightwalletd, and zebrad
- Removed `docker-ci/Dockerfile`, replacing it with a more structured CI setup

Improving CI setup structure by managing dependencies and build processes more efficiently in a new Dockerfile.

In the future, we might want to explore using `cargo make` for scripting the test command, for now, it requires a specific `docker run` command

nachog00 committed:
Add container image to CI jobs
- Updated `.github/workflows/ci.yml` to add `container` with `image: zingodevops/zaino-ci:latest` for `clippy` and `whitespace` jobs

Enhanced consistency in CI setup by using standardized container images for specified jobs.

nachog00 committed:
Remove Clippy output to GitHub step
- Deleted the 'Output clippy to Github' step from .github/workflows/ci.yml
- Simplified workflow by removing redundant Clippy output step

The Clippy output was deemed unnecessary for this setup, reducing clutter in the workflow.

nachog00 committed:
Add Docker container to Rustfmt job in CI workflow
- Updated `.github/workflows/ci.yml`
- Added `container` section with `image: zingodevops/zaino-ci:latest` in Rustfmt job

This change introduces containerization to ensure consistent environment during formatting step.

nachog00 committed:
Update cache key strategy in CI workflow (in check steps)

- Changed cache key from "rust-cache" to "cargo-deps" in .github/workflows/ci.yml
- Utilized Cargo.lock to hash the cache key instead of using job or run ID
- Improved cache efficiency by ensuring keys are tied to changes in dependencies

nachog00 committed:
moved docker job to another file

nachog00 committed:
Add condition checks for job execution in workflows
- Modified `.github/workflows/docker.yaml`
- Added a conditional check to ensure Docker image building only on merge (`docker` job)
- Added additional condition for release creation to only trigger after pull request merge (`create-release` job)

nachog00 committed:
Merge branch 'dev' into test-sh-cache

nachog00 committed:
Revert "on.push.branches= [test-sh-cache]"

This reverts commit 4c98a2d120da193b7f06ef023402f0ee02b17192.

nachog00 committed:
Update dependencies and streamline Docker setup
- Updated `.env.ci-dep-versions` to change `ZCASH_VERSION` from 6.1.0 to 6.2.0
- Modified `.gitignore` to include `docker_cargo`
- Simplified `Dockerfile.ci` by removing manual build steps for `zcashd` and replaced with a binary pull
- Adjusted `USER` and `GID` ARGs in `Dockerfile.ci` to match the typical user and avoid permission conflicts
- Added Rust toolchain installation in `Dockerfile.ci`
- Introduced new script `ci/build-ci-image.sh` for CI Docker image build
- Added `ci/test.sh` script for running Dockerized tests

This improves maintainability by automating dependency management and simplifying Docker configuration.

nachog00 committed:
Update Docker image in test script
- Changed Docker image tag from `zaino-ci:local` to `zingodevops/zaino-ci:latest` in `ci/test.sh`
- Ensures using the latest CI environment for testing

nachog00 committed:
Add CI/CD workflow for Docker image

- Created `.github/workflows/build-n-push-ci-image.yaml` for Docker image CI/CD
- Workflow is triggered on push to `dev` branch if `.env.ci-dep-versions` file changes
- Extracted version variables from `.env.zaino-versions` to use in Docker build
- Configured Docker image build and push, supporting multi-platform builds and version tagging

nachog00 committed:
Update .gitignore and add .gitkeep files
- Modified .gitignore to include docker_cargo/**/* for ignoring all files
- Added exception for docker_cargo/**/.gitkeep to retain empty directories
- Created .gitkeep in docker_cargo, docker_cargo/git, and docker_cargo/registry to ensure directories are tracked

`docker_cargo` is the container's .cargo
I tried having it share the host's .cargo, but faced many permission issues.

nachog00 committed:
Update CI workflow container images

- Changed container images in .github/workflows/ci.yml to versions with ':root' tag
- Ensures correct base image is utilized for CI processes

- This is to be corrected in the future to use a rootless image

nachog00 committed:
Switch to root user environment in Dockerfile and test script
- Updated `Dockerfile.ci` to use root user instead of creating a non-root user
- Commented out creation and setup of non-root user in `Dockerfile.ci`
- Modified `USER` and `HOME` arguments to correspond to root
- Altered `ci/test.sh` to use root image for testing purposes

Switching to root user simplifies permissions management and aligns container operations with root-based processes.

PD: this change is temporary, we expect to wind back to rootless when the conflict with self hosted runners is solved.

nachog00 committed:
Update Dockerfile.ci with version arguments and improved run usage
- Added RUST_VERSION, GO_VERSION, LIGHTWALLETD_VERSION, ZCASH_VERSION, and ZEBRA_VERSION arguments to final stage
- Updated LABEL usage instruction for improved run command with volume mounts

Enhance Docker image definition for better version control and usability in CI

nachog00 committed:
Fix CI workflow container images
- Changed container image from `zingodevops/zaino-ci:latest:root` to `zingodevops/zaino-ci:root` in ci.yml

dorianvp committed:
fix(ZcashIndexerRpcServer): register `getblockcount` method

dorianvp committed:
Merge branch 'dev' into fix_rpc_response

zancas commented:
@ala-mode any updates on this?  It looks like it hasn't been mentioned in 3 days?

dorianvp committed:
test(chain_cache): ignore tests involving `cache` & remove generation docs

ala-mode committed:
remove trait and move method

ala-mode committed:
moved method and removed impl

ala-mode committed:
added method to backend fetch

ala-mode committed:
add fn to jsonrpc service, adjust tags

ala-mode committed:
add struct and impl to response

ala-mode committed:
rename struct

ala-mode committed:
add fn to connector

ala-mode committed:
removed trait import from state backend

ala-mode committed:
add impl to server, add Hash type to indexer

ala-mode commented:
The consideration for a `BlockExplorerIndexer` trait has been factored out to be considered separately (#348).

idky137 committed:
Merge pull request #347 from zingolabs/ignore-tests-with-cache

test(chain_cache): ignore tests involving `cache`

idky137 committed:
Merge branch 'dev' into fix_rpc_response

idky137 committed:
Merge pull request #338 from AloeareV/fix_rpc_response

better zebra error propagation

idky137 committed:
clippy fixes

idky137 committed:
Merge branch 'dev' into finalised_state_2_0

nachog00 committed:
Merge branch 'dev' into test-sh-cache

AloeareV committed:
Merge pull request #344 from nachog00/test-sh-cache

ci still fails...but it fails more successfully. Great!

nachog00 committed:
Update Rust version in Dockerfile.ci
- Changed RUST_VERSION from 1.76.0 to 1.87.0

nachog00 committed:
Add script for copying and renaming binaries
- Created `ci/copy-bins.sh` to automate binary extraction
- Used Docker to copy binaries from `zaino-ci` image

This is a temporary workaround. Since some tests look for bins in PATH and some in /test_binaries/bins, we have to support both.
Issue is, test_binaries lives in the rootdir which is mounted, hence overwriten.
One solution could be an entrypoint script that copies them over... But we want to move away from that pattern anyways (the one of having bins in test_binaries)

nachog00 committed:
Remove lightwalletd from CI setup
- Removed GO_VERSION and LIGHTWALLETD_VERSION from .env.ci-dep-versions
- Removed ARG for GO_VERSION and LIGHTWALLETD_VERSION from Dockerfile.ci
- Deleted the lightwalletd build stage in Dockerfile.ci
- Removed COPY step for lightwalletd in Dockerfile.ci

Lightwalletd not needed on CI setup anymore

nachog00 committed:
Merge branch 'ci-image-pipeline-root' into ci-image-pipeline-zcashd-registry

nachog00 committed:
Revert "Merge branch 'ci-image-pipeline-root' into ci-image-pipeline-zcashd-registry"

This reverts commit 170218d883f443610e7cb05928aaaa12fb66500a, reversing
changes made to fffcfeb36b22a3cb9d61c5a02a9e34cc68f550c7.

nachog00 committed:
Updated Dockerfile.ci to use root user
- Changed ARG USER and ARG HOME to default to root
- Commented out non-root user setup section

Switched to root user for simplicity and to potentially resolve permission issues during CI builds.

nachog00 committed:
Remove GO and LIGHTWALLETD references in build script
- Removed echo statements for GO_VERSION and LIGHTWALLETD_VERSION from build-ci-image.sh
- Removed build arguments for GO_VERSION and LIGHTWALLETD_VERSION from docker build process in build-ci-image.sh

The GO and LIGHTWALLETD components are no longer required for the Docker image build process, thus their removal simplifies the script and reduces unnecessary complexity.

nachog00 committed:
Remove unused Go and LightwalletD versions from CI
- Modified build-n-push-ci-image.yaml to remove echo of GO_VERSION and LIGHTWALLETD_VERSION
- Removed LIGHTWALLETD_VERSION from image tags
- Deleted GO_VERSION and LIGHTWALLETD_VERSION from build arguments

The Go and LightwalletD versions are no longer relevant, reducing complexity in CI configuration.

AloeareV committed:
Merge pull request #345 from nachog00/ci-image-pipeline-zcashd-registry

improved ci image + deploy pipeline

dorianvp committed:
test(get_block_count): add remainin e2e tests

dorianvp committed:
Merge branch 'dev' into implement-getblockcount

ala-mode committed:
update comment label

idky137 committed:
Merge branch 'dev' into finalised_state_2_0

dorianvp committed:
Merge pull request #352 from ala-mode/implement_getblockcount

update comment label

AloeareV committed:
cargo fmt

AloeareV committed:
strip whitespace

AloeareV committed:
Merge pull request #330 from idky137/finalised_state_2_0

ChainBlock

dorianvp committed:
fix(Height): impl `From<GetBlockCountResponse>` instead of `TryFrom`

dorianvp committed:
Merge branch 'dev' into implement-getblockcount

nachog00 committed:
Exclude failing tests in CI due to zebra error
- Modified .config/nextest.toml to update the default-filter
- Added conditions to exclude specific tests related to zebra issues
- The changes are made to prevent CI failures until zebra releases a fix

zancas committed:
Merge pull request #354 from zingolabs/ignore-zebra-release-dependant-tests

Exclude failing tests in CI

ala-mode committed:
correct zaino fetch comments

dannywillems committed:
Update Dockerfile

ala-mode committed:
adjust comments

ala-mode committed:
Merge branch 'dev' into implement-getblockcount

zancas committed:
Merge pull request #357 from dannywillems/dannywillems-patch-1

Dockerfile: correctly support NO_TLS build argument

dorianvp committed:
Merge branch 'dev' into implement-getblockcount

nachog00 committed:
Fix trailing whitespace in nextest.toml
- Removed trailing whitespace from excluded tests in .config/nextest.toml

To improve code formatting and prevent potential issues with whitespace-sensitive tools.

AloeareV committed:
Merge pull request #358 from zingolabs/fix-trailing-whitespace-nextest-toml

Fix trailing whitespace in nextest.toml

dorianvp committed:
Merge branch 'dev' into implement-getblockcount

idky137 committed:
Merge pull request #359 from zingolabs/zaino-fetch_docfix

Zaino fetch docfix

idky137 committed:
Merge branch 'dev' into implement-getblockcount

idky137 committed:
Merge pull request #340 from zingolabs/implement-getblockcount

Implement `getblockcount`

AloeareV committed:
fix cargo update and update

AloeareV committed:
address todo comment

AloeareV committed:
add chaintip update subscriber

AloeareV committed:
add state_service chaintip_update_subscriber

AloeareV committed:
remove unneeded mut

AloeareV committed:
fix incorrect byte reversal in test

idky137 committed:
Merge pull request #363 from AloeareV/cargo_update

fix cargo update and update

idky137 committed:
Merge branch 'dev' into chaintip_update_subscriber

idky137 committed:
Merge pull request #349 from zingolabs/ci-image-pipeline-zcashd-registry

lightwalletd ci pipeline cleanup

idky137 committed:
Merge pull request #364 from zingolabs/chaintip_update_subscriber

Chaintip update subscriber

idky137 committed:
Merge branch 'dev' of github.com:zingolabs/zaino into feat-env-conf

idky137 committed:
fixed cargo.lock conflicts

idky137 committed:
Merge pull request #335 from gustavovalverde/feat-env-conf

refactor(config)!: migrate to `figment` for layered configuration parsing

ala-mode committed:
merge in dev

ala-mode committed:
change result to zebra_chain::block::Hash

ala-mode committed:
remove error type, return hash

ala-mode commented:
Closed in favor of #367
